File: cliphook.pl
:- use_module(library(process)).
:- use_module(library(lists)).
:- use_module(library(date)).
:- use_module(library(readutil)).

log_file('/home/dein_user/clipboard.log').
interval(2). % Sekunden

% --- Hash Funktion: einfache Summierung der Codes
hash_string(String, Hash) :-
    string_codes(String, Codes),
    sum_list(Codes, Hash).

% --- Clipboard auslesen
get_clipboard(Content) :-
    process_create(path(wl-paste), ['-n'], [stdout(pipe(Out))]),
    read_string(Out, _, ContentRaw),
    string_trim(ContentRaw, Content).

string_trim(Str, Trimmed) :-
    string_codes(Str, Codes),
    exclude(is_newline, Codes, TrimmedCodes),
    string_codes(Trimmed, TrimmedCodes).

is_newline(10). % '\n'
is_newline(13). % '\r'

% --- Clipboard loggen
log_clipboard(Content) :-
    get_time(TS),
    stamp_date_time(TS, DateTime, local),
    format_time(atom(TSString), '%Y-%m-%d %H:%M:%S', DateTime),
    log_file(File),
    open(File, append, Stream),
    format(Stream, "[~w]~n", [TSString]),
    format(Stream, "~w~n~n", [Content]),
    close(Stream).

% --- Hauptloop
clipboard_loop(LastHash) :-
    get_clipboard(Content),
    (Content \= "" ->
        hash_string(Content, CurrentHash),
        (CurrentHash \= LastHash ->
            log_clipboard(Content),
            sub_string(Content, 0, 30, _, Preview),
            format("Neuer Clipboard-Inhalt geloggt: ~w...~n", [Preview]),
            true
        ; true),
        sleep(interval(_)),
        clipboard_loop(CurrentHash)
    ; sleep(interval(_)),
      clipboard_loop(LastHash)
    ).

% --- Start
start :-
    clipboard_loop(0).

File: cliphook.zig
const std = @import("std");

const LOG_FILE = "/home/dein_user/clipboard.log";
const INTERVAL_MS = 2000;

fn getClipboard(allocator: *std.mem.Allocator) ![]u8 {
    var gpa = std.process.Process.spawn(.{
        .argv = &[_][]const u8{"wl-paste", "-n"},
        .stdout_behavior = .Capture,
    }) catch return error("Failed to spawn wl-paste");

    const result = gpa.stdout.readToEndAlloc(allocator) catch return error("Failed to read stdout");
    try gpa.wait();
    return std.mem.trimRight(result, "\n\r");
}

fn sha256Hash(allocator: *std.mem.Allocator, input: []const u8) ![]u8 {
    var hasher = std.crypto.hash.sha256.init();
    hasher.update(input);
    const hash = hasher.final(allocator) catch return error("SHA256 failed");
    return hash;
}

fn logClipboard(content: []const u8) !void {
    const now = std.time.milliTimestamp();
    var file = try std.fs.cwd().openFile(LOG_FILE, .{ .write = true, .append = true, .create = true });
    defer file.close();

    try file.writeAll("[" ++ std.fmt.allocPrint("{}\n", .{now}) catch "timestamp error");
    try file.writeAll(content);
    try file.writeAll("\n\n");
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    var lastHash: []u8 = &[_]u8{0} ** 32; // initial dummy hash

    while (true) {
        const content = try getClipboard(allocator);
        if (content.len != 0) {
            const currentHash = try sha256Hash(allocator, content);
            if (!std.mem.eql(u8, lastHash, currentHash)) {
                try logClipboard(content);
                std.debug.print("Neuer Clipboard-Inhalt geloggt: {s}...\n", .{std.mem.slice(content, 0, std.math.min(content.len, 30))});
                allocator.free(lastHash);
                lastHash = currentHash;
            } else {
                allocator.free(currentHash);
            }
        }
        std.time.sleep(std.time.millisecond * INTERVAL_MS);
    }
}

fn error(msg: []const u8) !void {
    std.debug.print("Fehler: {s}\n", .{msg});
    return error.Unexpected;
}

